File: auth-backend/infra/database/mappers/permission_mapper.py
"""
Permission Mapper
Maps between Permission domain model and PermissionModel database model
"""
from typing import Optional
import uuid
from core.domain.auth.permission import Permission, PermissionAction
from infra.database.models.permission import PermissionModel, PermissionActionDB


class PermissionMapper:
    """Mapper for Permission domain <-> database model"""
    
    @staticmethod
    def to_domain(model: PermissionModel) -> Permission:
        """Convert database model to domain model"""
        if model is None:
            return None
        
        return Permission(
            id=model.id,
            user_id=model.user_id,
            client_id=model.client_id,
            resource_type=model.resource_type,  # String direto! ✨
            resource_id=model.resource_id,
            action=PermissionAction(model.action.value),
            granted_by=model.granted_by,
            created_at=model.created_at
        )
    
    @staticmethod
    def to_model(domain: Permission) -> PermissionModel:
        """Convert domain model to database model"""
        if domain is None:
            return None
        
        # Generate ID if not present
        if not domain.id:
            domain.id = str(uuid.uuid4())
        
        return PermissionModel(
            id=domain.id,
            user_id=domain.user_id,
            client_id=domain.client_id,
            resource_type=domain.resource_type,  # String direto! ✨
            resource_id=domain.resource_id,
            action=PermissionActionDB(domain.action.value),
            granted_by=domain.granted_by,
            created_at=domain.created_at
        )

File: auth-frontend/src/infra/api/mappers/permission.mapper.ts
/**
 * Permission Mapper
 * Converts between DTOs and Domain models
 */

import {
  Permission,
  PermissionAction,
  PermissionListResponse,
  GrantPermissionRequest,
} from '../../../core/domain/permission';
import {
  PermissionDTO,
  GrantPermissionDTO,
  PermissionListResponseDTO,
} from '../dtos/permission.dto';

export class PermissionMapper {
  /**
   * Convert DTO to Domain model
   */
  static toDomain(dto: PermissionDTO): Permission {
    return {
      id: dto.id,
      user_id: dto.user_id,
      client_id: dto.client_id,
      resource_type: dto.resource_type,
      resource_id: dto.resource_id,
      action: dto.action as PermissionAction,
      granted_by: dto.granted_by,
      created_at: new Date(dto.created_at),
    };
  }

  /**
   * Convert Domain model to DTO
   */
  static toDTO(permission: Permission): PermissionDTO {
    return {
      id: permission.id,
      user_id: permission.user_id,
      client_id: permission.client_id,
      resource_type: permission.resource_type,
      resource_id: permission.resource_id,
      action: permission.action,
      granted_by: permission.granted_by,
      created_at: permission.created_at.toISOString(),
    };
  }

  /**
   * Convert list response DTO to Domain
   */
  static listToDomain(dto: PermissionListResponseDTO): PermissionListResponse {
    return {
      permissions: dto.permissions.map(this.toDomain),
      total: dto.total,
    };
  }

  /**
   * Convert grant request to DTO
   */
  static grantRequestToDTO(request: GrantPermissionRequest): GrantPermissionDTO {
    return {
      user_id: request.user_id,
      resource_type: request.resource_type,
      action: request.action,
      resource_id: request.resource_id,
    };
  }
}

